staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample-and-count.go

```diff
@@ -0,0 +1,193 @@
+		}}
+}
+
+func (swg *sampleAndCountObserverGenerator) Metrics() Registerables {
+	return Registerables{swg.samples, swg.counts}
+}
+
+type sampleAndCountHistograms struct {
+	*sampleAndCountObserverGenerator
+	labelValues      []string
+	countLabelValues [][]string // one []string for each bucket
+
+	sync.Mutex
+	denominator float64
+	sampleAndCountAccumulator
+}
+
+type sampleAndCountAccumulator struct {
+	lastSet    time.Time
+	lastSetInt int64 // lastSet / samplePeriod
+	numerator  float64
+	ratio      float64 // numerator/denominator
+	lastBucket int     // negative when not in a bucket
+}
+
+var _ RatioedChangeObserver = (*sampleAndCountHistograms)(nil)
+
+func (saw *sampleAndCountHistograms) Add(deltaNumerator float64) {
+	saw.innerSet(func() {
+		saw.numerator += deltaNumerator
+	})
+}
+
+func (saw *sampleAndCountHistograms) Observe(numerator float64) {
+	saw.innerSet(func() {
+		saw.numerator = numerator
+	})
+}
+
+func (saw *sampleAndCountHistograms) SetDenominator(denominator float64) {
+	saw.innerSet(func() {
+		saw.denominator = denominator
+	})
+}
+
+func findBucket(countBuckets []float64, x float64) int {
+	if x < countBuckets[0] {
+		return -1
+	}
+	var idx int
+	for idx = len(countBuckets) - 1; idx > 0 && x < countBuckets[idx]; idx-- {
+	}
+	return idx
+}
+
+func (saw *sampleAndCountHistograms) innerSet(updateNumeratorOrDenominator func()) {
+	when, whenInt, acc, wellOrdered := func() (time.Time, int64, sampleAndCountAccumulator, bool) {
+		saw.Lock()
+		defer saw.Unlock()
+		// Moved these variables here to tiptoe around https://github.com/golang/go/issues/43570 for #97685
+		when := saw.clock.Now()
+		whenInt := saw.quantize(when)
+		acc := saw.sampleAndCountAccumulator
+		dt := when.Sub(acc.lastSet)
+		wellOrdered := dt >= 0
+		updateNumeratorOrDenominator()
+		saw.ratio = saw.numerator / saw.denominator
+		if wellOrdered {
+			bucket := findBucket(saw.countBuckets, saw.ratio)
+			if saw.lastBucket >= 0 {
+				saw.counts.WithLabelValues(saw.countLabelValues[saw.lastBucket]...).Add(dt.Seconds())
+			}
+			saw.lastSet = when
+			saw.lastBucket = bucket
+		}
+		// `wellOrdered` should always be true because we are using
+		// monotonic clock readings and they never go backwards.  Yet
+		// very small backwards steps (under 1 microsecond) have been
+		// observed
+		// (https://github.com/kubernetes/kubernetes/issues/96459).
+		// In the backwards case, treat the current reading as if it
+		// had occurred at time `saw.lastSet` and log an error.  It
+		// would be wrong to update `saw.lastSet` in this case because
+		// that plants a time bomb for future updates to
+		// `saw.lastSetInt`.
+		return when, whenInt, acc, wellOrdered
+	}()
+	if !wellOrdered {
+		lastSetS := acc.lastSet.String()
+		whenS := when.String()
+		klog.Errorf("Time went backwards from %s to %s for labelValues=%#+v", lastSetS, whenS, saw.labelValues)
+	}
+	for acc.lastSetInt < whenInt {
+		saw.samples.WithLabelValues(saw.labelValues...).Observe(acc.ratio)
```

Comment:
I'm still struggling with this one a bit. Namely: what this metric really gives us.

Once we have a counter, we know how much time we actually spent in each of predefined buckets. So I don't really see how I would be supposed to use this histogram in addition to the counter above.

Reply:
But - the bigger the sample period we take, the less usable that is (as it may be completely inaccurate).
And additionally, we're not really solving the core problem, because we're still reporting a bunch of metrics here.

Also - I know that it gives us complete set of buckets - but how will I use it. What do I get from knowing that I spend 20% of time in 0.2 bucket instead of 0.3 bucket?
I guess my point is - if we add 0.5 in our counter (and maybe one more small value like 0.1 or sth) I don't know how I would ever want to use this metric.